src: add process.release.lts property & LTS CODENAME VOTING

rvagg opened the pull request:
Looking for review of this against `v4.x`, and separately calling for votes from @nodejs/collaborators on what the codename for our first LTS line (v4.x) will be. We are aiming for v4.2.0 LTS to be released this Thursday so this has to be quick!

Adopting a codename for LTS releases is part of the original LTS plan and although we've had a couple of late-breaking objections to even having a codename we're going ahead with it anyway given that this has been baked into the LTS plan that has already been adopted. We'll give it a go for this first one and assess for v6 if we go ahead and do it again.

The objectives are something like this:
* Something fun
* Something to set the LTS releases apart from the standard releases, both technically (a `process.release.lts` property) and for non-code communication purposes. We have an additional burden of having to start an LTS at an arbitrary point of a semver-major + semver-minor, i.e. 4.2.0 in this case, being able to call it _something_ from that point on sets it apart
* Foster a shared culture around LTS as more than just another release line, having input on what the codename should be from the broader group also helps with a sense of shared ownership (also, _fun_ again, imagine how boring Ubuntu would be without the fun names? It'd be as boring as Fedora has become)

The plan was for the @nodejs/lts group to come up with a short-list and call for voting. We've decided to go with the periodic-table for inspiration, there's a nice connection between the elements and the elemental nature of what we are trying to do with Node (i.e. small-core), also, there's lots to choose from! We also agreed that using alphabetic ordering will be helpful in deciphering order without needing additional context if people choose to refer to a line by its codename. These are what are available:

* Actinium Ac
* Aluminum Al
* Americium Am
* Antimony Sb
* Argon Ar
* Arsenic As
* Astatine At

For various reasons we've had to trim that list right down and only have two candidates for the v4 LTS codename:

### Actinium

Actinium is a silvery radioactive metallic element. Actinium glows in the dark due to its intense radioactivity with a blue light. Actinium was discovered in 1899 by André-Louis Debierne, a French chemist, who separated it from pitchblende. Friedrich Otto Giesel independently discovered actinium in 1902. The chemical behavior of actinium is similar to that of the rare earth lanthanum. The word actinium comes from the Greek aktis, aktinos, meaning beam or ray. Also, Actinium happens to have **4** isotopes, which is nice for our v4 release.

### Argon

Argon was suspected to be present in air by Henry Cavendish in 1785 but wasn't discovered until 1894 by Lord Rayleigh and Sir William Ramsay. Argon is the third noble gas, in period 8, and it makes up about 1% of the Earth's atmosphere. Argon has approximately the same solubility as oxygen and it is 2.5 times as soluble in water as nitrogen . This chemically inert element is colorless and odorless in both its liquid and gaseous forms. It is not found in any compounds. This gas is isolated through liquid air fractionation since the atmosphere contains only 0.94% argon. The Martian atmosphere in contrast contains 1.6% of Ar-40 and 5 ppm Ar-36. World production exceeds 750.000 tonnes per year, the supply is virtually inexhaustible.

(Astatine might also have been in the running except for the slightly awkward pronunciation and this gem: "total world production of astatine to date is estimated to be less than a millionth of a gram, and virtually all of this has now decayed away.")

So, this vote is open to @nodejs/collaborators only, you get one vote each, just leave a comment with your choice below.

Also looking for reviews of the changes this introduces (assuming the codename will be inserted to the appropriate places when chosen).

cjihrig commented:
Changes LGTM, once the real value is in there. My vote is for Argon because it's shorter, and easier to remember and say.

evanlucas commented:
+1 to Argon

Trott commented:
Another vote for Argon
On Tue, Oct 6, 2015 at 6:07 AM Evan Lucas <notifications@github.com> wrote:

> +1 to Argon
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3212#issuecomment-145851259>.
>

mhdawson commented:
Changes lgtm once real value is there.

+1 to Argon

thefourtheye commented on src/node.cc:
```diff
@@ -2736,6 +2736,9 @@ void SetupProcessObject(Environment* env,
   Local<Object> release = Object::New(env->isolate());
   READONLY_PROPERTY(process, "release", release);
   READONLY_PROPERTY(release, "name", OneByteString(env->isolate(), "node"));
+  READONLY_PROPERTY(release,
```
How does this make sure that `lts` is defined only for LTS releases?

rvagg replied:
this PR is against `v4.x`, that's how; I'm open to suggestions on doc changes that could go in to `master` if any

aheckmann commented:
+1 Argon

targos commented:
+1 for Argon, though I would have preferred if we picked the elements in the natural order and start with Hydrogen.

thefourtheye commented:
LGTM + Argon

jasnell commented:
+1 for Argon. Patch LGTM.

bnoordhuis commented:
Can we make `process.release.lts` a number or number-as-string?  Makes lexicographic comparison easy:

```javascript
if (process.release.lts < 201709) {  // do this one thing
if (process.release.lts > 201510) {  // do this other thing
```

ChALkeR commented:
@rvagg What will the 10th LTS release be called?

trevnorris commented:
+1 to Ben's suggestion.

Fishrock123 commented:
(we already bikeshedded that in the meeting)

Let's just do `{ name: 'Argon', date: 201510 }`

Trott commented:
![Actinium Defeats Argon](http://i.imgur.com/jDTMTsP.jpg)

rvagg commented:
@nodejs/lts opinions on @Fishrock123's suggestion above, my recollection is that we ended at the date not being all that helpful given that we already have `process.version` to tell us what version its running on. I'm -0 on making it a complex object.

I guess we could review the hangout video ... someone have the patience for that?

Fishrock123 commented:
> > Let's just do `{ name: 'Argon', date: 201510 }`

> @nodejs/lts opinions on @Fishrock123's suggestion above, my recollection is that we ended at the date not being all that helpful given that we already have `process.version` to tell us what version its running on. I'm -0 on making it a complex object.

Actually, same. Going back to @zkat and @issacs's points, this is what the semver version is for. We only have an lts bit in to tell that it is actually an lts release and whether or not we make it a string, it might as well be a boolean.

Fishrock123 commented:
Also, interesting thought: I guess this means we always have to bump a minor when going into an LTS phase?

jbergstroem commented:
Whatever name, just get @bnoordhuis's date parsing in there +1.

Fishrock123 commented:
> Whatever name, just get @bnoordhuis's date parsing in there +1.

As we bikeshedded in the LTS meeting, that's not actually useful. Just use the semver version instead (+ a check if it is an lts at all).

jbergstroem commented:
@Fishrock123 haven't caught up on the bikeshedding just yet. Thanks for pointing me though.

Fishrock123 commented:
Also, fwiw, +1 Argon (my suggestion)(am I allowed to vote for my own?)

rvagg commented:
Updated PR to use `Argon` and use node_version.h macros as per @jasnell's proposal.

Also added in more strict testing so that it'll enforce no `process.release.lts` unless `~4.2.0` in which case it must be `Argon`.

jasnell commented:
LGTM. There are a few bits I'll fix up when it lands (i.e. "WHAT WILL THIS BE?" is the doc change)

jasnell commented:
Landed in 42b936e
